cpp/src/groupby/sort/group_quantiles.cu

```diff
@@ -14,26 +14,64 @@
  * limitations under the License.
  */
 
-#include "group_reductions.hpp"
-
-#include <quantiles/quantiles_util.hpp>
-
 #include <cudf/column/column_device_view.cuh>
 #include <cudf/column/column_factories.hpp>
 #include <cudf/column/column_view.hpp>
 #include <cudf/detail/aggregation/aggregation.hpp>
-#include <cudf/types.hpp>
+#include <cudf/dictionary/detail/iterator.cuh>
+#include <cudf/dictionary/dictionary_column_view.hpp>
 
+#include <groupby/sort/group_reductions.hpp>
+#include <quantiles/quantiles_util.hpp>
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/device_vector.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <rmm/thrust_rmm_allocator.h>
 #include <thrust/for_each.h>
 
 namespace cudf {
 namespace groupby {
 namespace detail {
 namespace {
+
+template <typename ResultType, typename Iterator>
+struct calculate_quantile_fn {
+  Iterator values_iter;
+  column_device_view d_group_size;
+  mutable_column_device_view d_result;
+  size_type const* d_group_offset;
+  double const* d_quantiles;
+  size_type num_quantiles;
+  interpolation interpolation;
+
+  __device__ void operator()(size_type i)
+  {
+    size_type segment_size = d_group_size.element<size_type>(i);
+
+    auto d_itr = values_iter + d_group_offset[i];
+    thrust::transform(thrust::seq,
+                      d_quantiles,
+                      d_quantiles + num_quantiles,
+                      d_result.data<ResultType>() + i * num_quantiles,
+                      [d_itr, segment_size, interpolation = interpolation](auto q) {
+                        return cudf::detail::select_quantile_data<ResultType>(
+                          d_itr, segment_size, q, interpolation);
+                      });
+
+    size_type offset = i * num_quantiles;
+    thrust::for_each_n(thrust::seq,
+                       thrust::make_counting_iterator(0),
+                       num_quantiles,
+                       [d_result = d_result, segment_size, offset](size_type j) {
+                         if (segment_size == 0)
+                           d_result.set_null(offset + j);
+                         else
+                           d_result.set_valid(offset + j);
+                       });
```

Comment:
This looks like it will be very slow as it is setting each bit individually. Can't this use `valid_if`?

Reply:
Nevermind, I see we're inside a device function.